File: java/src/main/java/freetranslator/CortexScheduler.java
package freetranslator;

import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.*;
import java.util.stream.*;

/**
 * FreeTranslator v2.1 – Cortex Dinamikus Ütemező (Java port)
 *
 * GPU-analóg terhelés elosztás:
 *   Warp   = CortexUnit     (egy feldolgozó egység)
 *   Blokk  = önbeágyazott egységfa (mászó mélységnövekedés)
 *   Shader = feladattípus (É/K/D/Ny) – reticle szög alapján
 *
 * Csúszó  – load balancing: legkisebb terhelésű egység kap feladatot
 * Mászó   – önbeágyazás: túlterhelt egység gyermeket szül (depth++)
 * Forgó   – vektorirányzék: reticle szög → feladattípus (0/90/180/270°)
 *
 * Hőszabályzás:
 *   thermal = EMA(egységek heat átlaga) – valós időben frissül
 *   COOL/WARM/HOT/CRITICAL sávok → Ollama throttle szorzók
 *
 * Thread-safe: ConcurrentHashMap + AtomicInteger
 */
public class CortexScheduler {

    // ── Thermal sávok ────────────────────────────────────────────────────────

    public enum ThermalBand {
        COOL    ("HŰVÖS",    0.30, 1.00, 1.00, 0),
        WARM    ("MELEG",    0.60, 0.85, 0.90, 0),
        HOT     ("FORRÓ",    0.80, 0.65, 0.70, 80),
        CRITICAL("KRITIKUS", 1.00, 0.40, 0.50, 200);

        public final String  label;
        public final double  maxLevel;
        public final double  ctxMult;
        public final double  predictMult;
        public final int     delayMs;

        ThermalBand(String label, double max, double ctx, double pred, int delay) {
            this.label = label; this.maxLevel = max;
            this.ctxMult = ctx; this.predictMult = pred; this.delayMs = delay;
        }

        public static ThermalBand of(double level) {
            for (ThermalBand b : values()) if (level <= b.maxLevel) return b;
            return CRITICAL;
        }
    }

    // ── CortexUnit ───────────────────────────────────────────────────────────

    public static class CortexUnit {
        private static final AtomicInteger SEQ = new AtomicInteger(1);

        public final String id;
        public final String type;
        public final int    depth;
        public final String parentId;

        private volatile double load  = 0;
        private volatile double heat  = 0;
        private volatile int    tasks = 0;
        private volatile boolean active = true;
        private volatile long   lastTaskMs = 0;

        private final List<CortexUnit> children = new CopyOnWriteArrayList<>();

        CortexUnit(String type, int depth, String parentId) {
            this.id       = String.format("CU%03d", SEQ.getAndIncrement());
            this.type     = type;
            this.depth    = depth;
            this.parentId = parentId;
        }

        /** Csúszó terhelés felvétel */
        synchronized void acquire(double complexity) {
            load = Math.min(1, load + complexity);
            heat = Math.min(1, heat + complexity * 0.35);
            tasks++;
            lastTaskMs = System.currentTimeMillis();
        }

        /** Hő-elmaradásos felszabadítás */
        synchronized void release(double complexity, long durationMs) {
            load = Math.max(0, load - complexity);
            double coolRate = complexity * 0.12 * Math.min(2.0, 1000.0 / Math.max(50, durationMs));
            heat = Math.max(0, heat - coolRate);
        }

        /** Mászó önbeágyazás – gyermek unit szülése */
        CortexUnit embed(String childType) {
            CortexUnit child = new CortexUnit(childType, depth + 1, id);
            children.add(child);
            return child;
        }

        public double getLoad()   { return load; }
        public double getHeat()   { return heat; }
        public int    getTasks()  { return tasks; }
        public boolean isActive() { return active; }
        public void retire()      { active = false; }

        public Map<String, Object> toMap() {
            Map<String, Object> m = new LinkedHashMap<>();
            m.put("id", id); m.put("type", type); m.put("depth", depth);
            m.put("parentId", parentId);
            m.put("load",  Math.round(load  * 1000) / 1000.0);
            m.put("heat",  Math.round(heat  * 1000) / 1000.0);
            m.put("tasks", tasks); m.put("active", active);
            m.put("childCount", children.stream().filter(CortexUnit::isActive).count());
            return m;
        }
    }

    // ── Scheduler ────────────────────────────────────────────────────────────

    private static final List<String> TASK_TYPES = List.of("north","east","south","west");
    private static final Map<String, Integer> TYPE_ANGLES =
        Map.of("north",0,"east",90,"south",180,"west",270);

    private final Map<String, CortexUnit> units = new ConcurrentHashMap<>();
    private volatile double thermal = 0;
    private final List<Map<String, Object>> log = new CopyOnWriteArrayList<>();

    public CortexScheduler() {
        for (String t : TASK_TYPES) {
            CortexUnit u = new CortexUnit(t, 0, null);
            units.put(u.id, u);
        }
    }

    /** Forgó vektorirányzék: reticle szög → feladattípus */
    public String angleToType(double deg) {
        double a = ((deg % 360) + 360) % 360;
        return TASK_TYPES.stream().min(Comparator.comparingDouble(t -> {
            double diff = Math.abs(a - TYPE_ANGLES.get(t));
            return Math.min(diff, 360 - diff);
        })).orElse("east");
    }

    /**
     * Egység foglalás – csúszó load balancing + mászó önbeágyazás ha szükséges.
     * Ha nincs szabad (load ≤ 0.85) azonos típusú egység → önbeágyazás.
     */
    public CortexUnit allocate(String type, double complexity, Double reticleAngle) {
        String taskType = (reticleAngle != null) ? angleToType(reticleAngle) : (type != null ? type : "east");

        List<CortexUnit> free = units.values().stream()
            .filter(u -> u.isActive() && u.type.equals(taskType) && u.getLoad() <= 0.85)
            .sorted(Comparator.comparingDouble(CortexUnit::getLoad))
            .collect(Collectors.toList());

        CortexUnit unit;
        if (!free.isEmpty()) {
            unit = free.get(0);
        } else {
            // Mászó önbeágyazás
            Optional<CortexUnit> deepest = units.values().stream()
                .filter(u -> u.isActive() && u.type.equals(taskType))
                .max(Comparator.comparingInt(u -> u.depth));

            if (deepest.isPresent()) {
                unit = deepest.get().embed(taskType);
                units.put(unit.id, unit);
            } else {
                unit = units.values().stream()
                    .filter(CortexUnit::isActive)
                    .min(Comparator.comparingDouble(CortexUnit::getLoad))
                    .orElseThrow(() -> new IllegalStateException("No active cortex unit"));
            }
        }

        unit.acquire(complexity);
        tick(taskType, complexity, "acquire");
        return unit;
    }

    /** Feladat befejezése + hőszabályzó felszabadítás */
    public void release(String unitId, double complexity, long durationMs) {
        CortexUnit unit = units.get(unitId);
        if (unit == null) return;
        unit.release(complexity, durationMs);
        tick(unit.type, complexity, "release");
        if (unit.depth > 0 && unit.getLoad() < 0.04 && unit.getTasks() > 0) {
            unit.retire();
        }
    }

    private synchronized void tick(String type, double complexity, String phase) {
        List<CortexUnit> active = activeUnits();
        double avgHeat = active.stream().mapToDouble(CortexUnit::getHeat).average().orElse(0);
        thermal = thermal * 0.72 + avgHeat * 0.28;

        Map<String, Object> entry = new LinkedHashMap<>();
        entry.put("ts", System.currentTimeMillis());
        entry.put("type", type); entry.put("complexity", complexity); entry.put("phase", phase);
        log.add(entry);
        if (log.size() > 20) log.remove(0);
    }

    /** GPU-analóg Ollama paraméter throttling */
    public ThrottleResult throttleParams(int numCtx, int numPredict, double temperature) {
        ThermalBand band = ThermalBand.of(thermal);
        int   ctx  = Math.max(512,  (int)(numCtx     * band.ctxMult));
        int   pred = Math.max(64,   (int)(numPredict  * band.predictMult));
        double temp = band == ThermalBand.CRITICAL ? temperature * 1.15 : temperature;
        return new ThrottleResult(ctx, pred, Math.round(temp * 100.0) / 100.0, band);
    }

    public record ThrottleResult(int numCtx, int numPredict, double temperature, ThermalBand band) {
        public int delayMs() { return band.delayMs; }
    }

    public ThermalBand getThermalBand() { return ThermalBand.of(thermal); }
    public double getThermal()          { return thermal; }
    public int getThermalPct()          { return (int) Math.round(thermal * 100); }

    private List<CortexUnit> activeUnits() {
        return units.values().stream().filter(CortexUnit::isActive).collect(Collectors.toList());
    }

    public int getMaxDepth() {
        return activeUnits().stream().mapToInt(u -> u.depth).max().orElse(0);
    }

    /** Terhelés eloszlás irányonként */
    public Map<String, Map<String, Object>> getLoadDist() {
        Map<String, Map<String, Object>> dist = new LinkedHashMap<>();
        for (String t : TASK_TYPES) {
            List<CortexUnit> us = activeUnits().stream()
                .filter(u -> u.type.equals(t)).collect(Collectors.toList());
            double avg = us.stream().mapToDouble(CortexUnit::getLoad).average().orElse(0);
            Map<String, Object> entry = new LinkedHashMap<>();
            entry.put("count",   us.size());
            entry.put("avgLoad", Math.round(avg * 1000.0) / 1000.0);
            dist.put(t, entry);
        }
        return dist;
    }

    /** Teljes állapot snapshot */
    public Map<String, Object> status() {
        ThermalBand band   = ThermalBand.of(thermal);
        List<CortexUnit> active = activeUnits();
        Map<String, Object> s = new LinkedHashMap<>();
        s.put("thermal",    Math.round(thermal * 10000.0) / 10000.0);
        s.put("thermalPct", getThermalPct());
        s.put("bandName",   band.name());
        s.put("bandLabel",  band.label);
        s.put("ctxMult",    band.ctxMult);
        s.put("predictMult",band.predictMult);
        s.put("delayMs",    band.delayMs);
        s.put("unitCount",  active.size());
        s.put("maxDepth",   getMaxDepth());
        s.put("load",       getLoadDist());
        s.put("units",      active.stream().map(CortexUnit::toMap).collect(Collectors.toList()));
        s.put("log",        log.stream().skip(Math.max(0, log.size()-10)).collect(Collectors.toList()));
        return s;
    }
}

File: src/cortex-scheduler.js
/**
 * Cortex Dinamikus Ütemező – GPU-analóg terhelés elosztás
 *
 * Analógia: GPU warp / blokk / SM ütemező
 *   Warp   = CortexUnit  (egy feldolgozó egység, irányvektorra hangolt)
 *   Blokk  = önbeágyazott CortexUnit fa (mélység növekedhet)
 *   Shader = feladattípus (É/K/D/Ny) – reticle szög alapján
 *
 * Csúszó  – egységek fokozatosan átveszik egymás terhét (load slide)
 * Mászó   – beágyazott gyermek egységek lassan migrálnak fel/le (depth crawl)
 * Forgó   – reticle szög forgatja a feladattípus-hozzárendelést (vector rotate)
 *
 * Hőszabályzás:
 *   Thermal = EMA(egységek heat értékei)
 *   HŰVÖS < 0.30 → teljes kapacitás
 *   MELEG  0.30–0.60 → ctx×0.85, pred×0.90
 *   FORRÓ  0.60–0.80 → ctx×0.65, pred×0.70 + 80ms késleltetés
 *   KRITIKUS > 0.80  → ctx×0.40, pred×0.50 + 200ms késleltetés
 *
 * @module cortex-scheduler
 */

'use strict';

const TASK_TYPES  = ['north', 'east', 'south', 'west'];
const TASK_ANGLES = { north: 0, east: 90, south: 180, west: 270 };

const THERMAL_BANDS = [
  { name: 'COOL',     label: 'HŰVÖS',    max: 0.30, color: '#78c4ff', ctxMult: 1.00, predictMult: 1.00, delayMs: 0   },
  { name: 'WARM',     label: 'MELEG',    max: 0.60, color: '#8be28b', ctxMult: 0.85, predictMult: 0.90, delayMs: 0   },
  { name: 'HOT',      label: 'FORRÓ',    max: 0.80, color: '#ffc96b', ctxMult: 0.65, predictMult: 0.70, delayMs: 80  },
  { name: 'CRITICAL', label: 'KRITIKUS', max: 1.00, color: '#ff9b9b', ctxMult: 0.40, predictMult: 0.50, delayMs: 200 }
];

function getThermalBand(level) {
  return THERMAL_BANDS.find(b => level <= b.max) || THERMAL_BANDS[THERMAL_BANDS.length - 1];
}

// ── CortexUnit ────────────────────────────────────────────────────────────────

let _nextUnitId = 1;

class CortexUnit {
  constructor(type, depth = 0, parentId = null) {
    this.id       = `CU${String(_nextUnitId++).padStart(3, '0')}`;
    this.type     = type;
    this.depth    = depth;
    this.parentId = parentId;
    this.children = [];   // beágyazott gyermek egységek
    this.load     = 0;    // aktuális terhelés 0–1
    this.heat     = 0;    // akkumulált hő 0–1
    this.tasks    = 0;    // elvégzett feladatok száma
    this.active   = true;
    this.born     = Date.now();
    this.lastTask = null;
  }

  /** Feladat felvétele – csúszó terhelésnövelés */
  acquire(complexity) {
    this.load     = Math.min(1, this.load + complexity);
    this.heat     = Math.min(1, this.heat + complexity * 0.35);
    this.tasks++;
    this.lastTask = Date.now();
  }

  /** Feladat befejezése – hő elmaradással csökken (hőszabályzó hatás) */
  release(complexity, durationMs) {
    this.load = Math.max(0, this.load - complexity);
    // Rövid feladat kevésbé hűt – hő-kapacitás analóg
    const coolRate = complexity * 0.12 * Math.min(2, 1000 / Math.max(50, durationMs));
    this.heat = Math.max(0, this.heat - coolRate);
  }

  /** Önbeágyazás – gyermek CortexUnit szülése (mászó mélységnövekedés) */
  embed(type) {
    const child = new CortexUnit(type, this.depth + 1, this.id);
    this.children.push(child);
    return child;
  }

  toJSON() {
    return {
      id: this.id, type: this.type, depth: this.depth, parentId: this.parentId,
      load:  +this.load.toFixed(3),
      heat:  +this.heat.toFixed(3),
      tasks: this.tasks,
      active: this.active,
      childCount: this.children.filter(c => c.active).length
    };
  }
}

// ── CortexScheduler ───────────────────────────────────────────────────────────

class CortexScheduler {
  constructor() {
    this.units   = {};   // id → CortexUnit
    this.thermal = 0;    // rendszer hőmérséklet EMA
    this.log     = [];   // utolsó 20 feladat esemény

    // Alap négy egység – egy irányonként (warp-0 analóg)
    for (const t of TASK_TYPES) {
      const u = new CortexUnit(t, 0, null);
      this.units[u.id] = u;
    }
  }

  /** Reticle szög → feladattípus (forgó vektorirányzék routing) */
  angleToType(deg) {
    const a = ((deg % 360) + 360) % 360;
    return TASK_TYPES.reduce((best, t) => {
      const d  = Math.min(Math.abs(a - TASK_ANGLES[t]),   360 - Math.abs(a - TASK_ANGLES[t]));
      const bd = Math.min(Math.abs(a - TASK_ANGLES[best]), 360 - Math.abs(a - TASK_ANGLES[best]));
      return d < bd ? t : best;
    });
  }

  /**
   * Egység foglalása feladathoz.
   * Ha nincs szabad kapacitás (load > 0.85), önbeágyazás: új gyermek egység.
   */
  allocate(type, complexity = 0.5, reticleAngle = null) {
    const taskType = (reticleAngle != null) ? this.angleToType(reticleAngle) : (type || 'east');

    // Csúszó keresés: legkisebb terhelésű szabad egység azonos típusból
    const free = Object.values(this.units)
      .filter(u => u.active && u.type === taskType && u.load <= 0.85)
      .sort((a, b) => a.load - b.load);

    let unit;
    if (free.length > 0) {
      unit = free[0];
    } else {
      // Minden azonos típusú egység túlterhelt → mászó önbeágyazás
      const overloaded = Object.values(this.units)
        .filter(u => u.active && u.type === taskType)
        .sort((a, b) => b.depth - a.depth);

      if (overloaded.length > 0) {
        // Legmélyebb szinten lévő egység ágyazza be az újat
        unit = overloaded[0].embed(taskType);
        this.units[unit.id] = unit;
      } else {
        // Nincs ilyen típus → legenyhébb bármelyik
        unit = Object.values(this.units)
          .filter(u => u.active)
          .sort((a, b) => a.load - b.load)[0];
      }
    }

    unit.acquire(complexity);
    this._tick(taskType, complexity, 'acquire');
    return unit;
  }

  /** Feladat befejezése, egység hőjének visszaszabályzása */
  release(unitId, complexity = 0.5, durationMs = 500) {
    const unit = this.units[unitId];
    if (!unit) return;
    unit.release(complexity, durationMs);
    this._tick(unit.type, complexity, 'release');
    // Nem alap egység, nincs terhelés → visszavonul (hőszabályzó lehűlés)
    if (unit.depth > 0 && unit.load < 0.04 && unit.tasks > 0) {
      unit.active = false;
    }
  }

  /** EMA hőmérséklet frissítés */
  _tick(type, complexity, phase) {
    const active  = Object.values(this.units).filter(u => u.active);
    const avgHeat = active.reduce((s, u) => s + u.heat, 0) / Math.max(1, active.length);
    this.thermal  = +(this.thermal * 0.72 + avgHeat * 0.28).toFixed(4);

    this.log.push({ ts: Date.now(), type, complexity, phase });
    if (this.log.length > 20) this.log.shift();
  }

  getThermalBand() { return getThermalBand(this.thermal); }

  /**
   * GPU-analóg Ollama paraméter throttling.
   * Visszaadja a hőmérséklet-arányosan csökkentett paramétereket.
   */
  throttleParams(opts = {}) {
    const tb  = getThermalBand(this.thermal);
    const out = { ...opts };
    if (out.num_ctx)     out.num_ctx     = Math.max(512, Math.floor((out.num_ctx     || 2048) * tb.ctxMult));
    if (out.num_predict) out.num_predict = Math.max(64,  Math.floor((out.num_predict || 256)  * tb.predictMult));
    if (out.temperature) out.temperature = +(out.temperature * (tb.name === 'CRITICAL' ? 1.15 : 1.0)).toFixed(2);
    return { params: out, band: tb, delayMs: tb.delayMs };
  }

  /** Terhelés eloszlás irányonként */
  getLoadDist() {
    const dist = {};
    for (const t of TASK_TYPES) {
      const us = Object.values(this.units).filter(u => u.active && u.type === t);
      dist[t] = {
        count:   us.length,
        avgLoad: us.length ? +(us.reduce((s, u) => s + u.load, 0) / us.length).toFixed(3) : 0,
        avgHeat: us.length ? +(us.reduce((s, u) => s + u.heat, 0) / us.length).toFixed(3) : 0
      };
    }
    return dist;
  }

  getMaxDepth() {
    return Math.max(0, ...Object.values(this.units).filter(u => u.active).map(u => u.depth));
  }

  /** Teljes állapot snapshot – GET /api/cortex számára */
  status() {
    const active = Object.values(this.units).filter(u => u.active);
    const band   = getThermalBand(this.thermal);
    return {
      thermal:   this.thermal,
      thermalPct: Math.round(this.thermal * 100),
      band,
      unitCount: active.length,
      maxDepth:  this.getMaxDepth(),
      load:      this.getLoadDist(),
      units:     active.map(u => u.toJSON()),
      log:       this.log.slice(-10)
    };
  }
}

// Singleton – szerver szintű, megosztott állapot
const scheduler = new CortexScheduler();

module.exports = { CortexScheduler, CortexUnit, scheduler, getThermalBand, THERMAL_BANDS, TASK_TYPES };
